File: iosApp/iosApp/Modules/Root/RootPresenter.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol RootPresenter {
    func present()
}

final class DefaultRootPresenter {
    private weak var view: RootView?
    private let wireframe: RootWireframe
    private let signerStoreService: SignerStoreService

    init(
        view: RootView,
        wireframe: RootWireframe,
        signerStoreService: SignerStoreService
    ) {
        self.view = view
        self.wireframe = wireframe
        self.signerStoreService = signerStoreService
    }
}

extension DefaultRootPresenter: RootPresenter {

    func present() {
        wireframe.navigate(
            to: signerStoreService.items().isEmpty ? .keyStore : .dashboard,
            animated: false
        )
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/root/RootPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.root

import com.sonsofcrypto.web3lib.services.keyStore.SignerStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.modules.root.RootWireframeDestination.DASHBOARD
import com.sonsofcrypto.web3walletcore.modules.root.RootWireframeDestination.KEYSTORE

interface RootPresenter {
    fun present()
}

final class DefaultRootPresenter(
    private var rootView: WeakRef<RootView>,
    private val wireframe: RootWireframe,
    private val signerStoreService: SignerStoreService,
) : RootPresenter {

    override fun present() {
        wireframe.navigate(
            if(signerStoreService.items().isEmpty()) KEYSTORE else DASHBOARD
        )
    }
}
